### PR TITLE
Utils/graph

### PR DESCRIPTION
**Pull request by rkeeves**
## Overview:

This branch is about graph representation.

- **Graph interface** is the contract for graphs in general.
- **UndirectedGraph** extends **Graph** by undirected edge specific methods.
- **DirectedGraph** extends **Graph** by undirected edge specific methods.
- **HashGraph** is a naive implementation which supports both **UndirectedGraph** and **DirectedGraph**.
- **HashGraphTest** contains tests for **HashGraph** both as an **UndirectedGraph** and as a **DirectedGraph**.

## Compatibility

This branch has only additions.
No changes, no deletions.

This branch creates a new package called **hu/unideb/inf/utils**, because these classes/interfaces are more general,
so I figured it wouldn't make much sense shoving them into the **hu/unideb/inf/model** package.

Please feel free to critique or suggest changes.

### CODE REVIEW

File: src/main/java/hu/unideb/inf/utils/DirectedGraph.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package hu.unideb.inf.utils;
+
+import java.util.Set;
+
+/**
+ * DirectedGraph graphs is a kind of Graph.
+ * Edges connect two nodes.
+ * Directed means that A->B does not imply B->A.
+ * @param <N>
```
Comment (palmarcell) [inline src/main/java/hu/unideb/inf/utils/DirectedGraph.java:9]: Please add descriptions to the `@params` block throughout the pull request like:
`@param <N> represents a specific node class`

File: src/main/java/hu/unideb/inf/utils/DirectedGraph.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package hu.unideb.inf.utils;
```
Comment (palmarcell) [inline src/main/java/hu/unideb/inf/utils/DirectedGraph.java:1]: Please rename the "utils" package to a more concrete name.
It is a best practice to avoid using "util" package names, so we could separate classes and files based on their true meaning.
Read more about this here: [https://dave.cheney.net/2019/01/08/avoid-package-names-like-base-util-or-common](https://dave.cheney.net/2019/01/08/avoid-package-names-like-base-util-or-common)

File: src/main/java/hu/unideb/inf/utils/HashGraph.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        Objects.requireNonNull(target);
+        edges.add(new Edge(source, target));
+    }
+
+    @Override
+    public Set<N> getNodesReachableFrom(N src) {
+        return edges
+                .stream()
+                .filter(edge -> edge.getSource().equals(src))
+                .map(edge -> edge.getTarget()).collect(Collectors.toSet());
```
Comment (palmarcell) [inline src/main/java/hu/unideb/inf/utils/HashGraph.java:49]: It's really cool that you used stream here! :)
However, some minor modifications could be applied:
- "src" should be renamed to "source", so the naming would be consistent (if you take a look to the method above)
- You could put the `.collect()` part to a new line.

File: src/test/java/hu/unideb/inf/utils/HashGraphTest.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+import java.util.HashSet;
+import java.util.Set;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class HashGraphTest {
+
+    @Test
+    void given_graph_when_addDirectedEdgeAndSourceIsNull_Throw() {
```
Comment (palmarcell) [inline src/test/java/hu/unideb/inf/utils/HashGraphTest.java:14]: The naming should look like:
`methodName_StateUnderTest_ExpectedBehaviour`

So this method name would look like:
addDirectedEdge_WhenSourceNodeIsNull_ShouldThrowNullPointerException

Please rename the other test methods based on this structure.

File: src/test/java/hu/unideb/inf/utils/HashGraphTest.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        // given
+        HashGraph<Integer> g = new HashGraph<Integer>();
```
Comment (palmarcell) [inline src/test/java/hu/unideb/inf/utils/HashGraphTest.java:16]: Please use descriptive variable names.
I know "g" is pretty straightforward as a meaning of graph however, the code would be more readable if you would write out the full name of it.

**Review by palmarcell — CHANGES_REQUESTED** [review]

**Comment by rkeeves** [comment]
## Changes

### Additions

- added 4 missing test case for addUndirectedEdge to check null as param (throws java.lang.NullPointerException)
- added 1 missing test case for getNodesReachableFrom to check null as param does NOT throw java.lang.NullPointerException

### Modifications
- add description to @params blocks
- renamed utils package to datastructures
- renamed params (so naming would be consistent)
- formatted the .collect call to newline in long function chaining
- added missing javadocs for types and methods
- added notes to method javadocs where java.lang.NullPointerException can be thrown
- renamed all test cases according to methodName_StateUnderTest_ExpectedBehaviour
- renamed variables in test cases to be more descriptive instead of 'g'
- removed unnecessary inferables from tests ```HashSet<Integer> hashSet = new HashSet<>()```
- in tests changed Set creation to oneliner ```Set<Integer> expected = new HashSet<>(Collections.singletonList(1));```

## Possible, but not yet implemented changes

### 1. Should we allow function chaining?

Aka all 'add' functions currently return void.
We can instead return the graph instance.
Conflicts with idea 2.
  
### 2. Should we make it safe?

Aka currently all 'edge adding' methods throw on null param.
We can instead return a boolean false.
Conflicts with idea 1.

**Comment by palmarcell** [comment]
> ## Possible, but not yet implemented changes
> ### 1. Should we allow function chaining?
> Aka all 'add' functions currently return void.
> We can instead return the graph instance.
> Conflicts with idea 2.
> 
> ### 2. Should we make it safe?
> Aka currently all 'edge adding' methods throw on null param.
> We can instead return a boolean false.
> Conflicts with idea 1.

There's another option where we could pass a collection of the node pairs to a new method like `addDirectedEdges()` as parameters.
So the function could iterate over the pairs and call the already existing `addDirectedEdge()` method. This would require changing the access modifier on the Edge class so that we could use this new method like this:

```
hashGraph.addDirectedEdges(List.of(
        new Edge(firstNode, secondNode),
        new Edge(firstNode, thirdNode)
        ));
```

Throwing NullPointerException is a great practice, so that I wouldn't change that part.

However, we should think about that case, when the above-mentioned list contains an Edge instance with a null value:
```
hashGraph.addDirectedEdges(List.of(
        new Edge(firstNode, secondNode),
        new Edge(firstNode, null)
        ));
```

So my advice is that we should first check the parameter list, whether it's valid or not, and then we could start iterating over it.

What do you think about this?

**Comment by rkeeves** [comment]
## addDirectedEdges/addAllDirectedEdges

Your idea is actually really great when I think about possible use cases.
(I mean like reading room data from xml using JAXB for the game so we want to populate the graph with one short call because we have all the necessary data already from the xml.)

I have though two questions about the implementation of this approach:

### 1. Decorator Pattern

Maybe just for learning/demonstration purposes, these _addAllBlaBla_ methods could be provided by a decorator instead.
```
DirectedGraph<Integer> directedGraph = new HashGraph<>();
EnMasseDirectedGraph<Integer> enMasseDirectedGraph = new EnMasseDirectedGraph<>(directedGraph);
enMasseDirectedGraph.addAllDirectedEdges( ... );
```
My reasoning behind this is the following:
In the below case we must implement the same method twice in each of these classes:
```
EnMasseDirectedGraph<Integer> a = new HashGraph<>();
EnMasseDirectedGraph<Integer> b = new ArrayGraph<>();
```
But the logic will be the same, so we could do instead this:
```
EnMasseDirectedGraph<Integer> a = new EnMasseDirectedGraph<>(new HashGraph<>());
EnMasseDirectedGraph<Integer> b = new EnMasseDirectedGraph<>(new ArrayGraph<>());
```
So we only need to implement once this _addAll_ in _EnMasseDirectedGraph_.

#### The **downsides** of making this approach our API:
- makes the API not that straightforward, because you have to use the decorator instead of the concrete class for en masse additions
- makes it impossible for implementations _(like HashGraph and ArrayGraph)_ to optimize (aka knowing beforehand how many new edges you will possibly add can help greatly in an array based graph implementation for example)
- runs contrary to JDK standard (java.util classes support addAll without decorator)

#### The upside of making this approach optionally viable for use WITHIN IMPLEMENTATIONS:
Even if we decide that Decorator pattern should not be used for clients, it is still a valid option for internal use.
Consider the following scenario:
You make a new implementation for DirectedGraph called FooGraph.
If you don't want to/ can't optimize en masse addition in your _addAllDirectedEdges_ method, then you can simply do:
```
public class FooGraph<N> implements DirectedGraph<N>{

/* ... */

  @Override
  public void addAllDirectedEdges( /* args */ ){
    new EnMasseDirectedGraph<N>(this).addAllDirectedEdges(  /* args */  );
  }

}
```
Which results in less boilerplate.

### 2. Promote Edge into its own file

I might be paranoid, but exposing the Edge means that we're strongly coupling the caller to our implementation.

Imagine the following scenario:
Some student in our group decides to write an array based Graph implementation.
Then if he wants to support _addAll_ methods then he must use Edge which is currently an inner class of HashGraph.
If we decide that Edge will be exposed to clients for all graphs (instead of using [something else](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/tuple/Pair.html)) then Edge must be - I think - moved to its own file.

If we decide on that all graphs will use a standalone _Edge_ class to communicate with the callers, then **null check** for its fields can be moved to its **factory**, **factory method** or **ctor**.
(I think an Edge should never hold null values, because mathematically an Edge between two already known Nodes either exists or not. Aka, for me, there's no reason why we would ever want to create an Edge with any of its fields as null.)
